src/gap_to_julia.jl

```diff
@@ -64,269 +99,327 @@ julia> GAP.gap_to_julia( Matrix{Int}, val )
 ## Symbols
-gap_to_julia(::Type{Symbol}, obj::GapObj) = Symbol(obj)
+gap_to_julia_intern(::Type{Symbol}, obj::GapObj, recursive::Bool, dict::Union{Nothing,RecDict}) = Symbol(obj)
 
 ## Convert GAP string to Vector{UInt8} (==Vector{UInt8})
-function gap_to_julia(::Type{Vector{UInt8}}, obj::GapObj)
-    Wrappers.IsStringRep(obj) && return CSTR_STRING_AS_ARRAY(obj)
-    Wrappers.IsList(obj) && return UInt8[gap_to_julia(UInt8, obj[i]) for i = 1:length(obj)]
-    throw(ConversionError(obj, Vector{UInt8}))
+function gap_to_julia_intern(::Type{Vector{UInt8}}, obj::GapObj, recursive::Bool, dict::Union{Nothing,RecDict})
+    return _get!(dict, Vector{UInt8}, obj) do
+        Wrappers.IsStringRep(obj) && return CSTR_STRING_AS_ARRAY(obj)
+        Wrappers.IsList(obj) && return UInt8[UInt8(obj[i]) for i = 1:length(obj)]
+        throw(ConversionError(obj, Vector{UInt8}))
+    end
 end
 
-## BitVectors
-gap_to_julia(::Type{BitVector}, obj::GapObj) = BitVector(obj)
+gap_to_julia_intern(::Type{BitVector}, obj::GapObj, recursive::Bool, dict::Union{Nothing,RecDict}) = _get!(dict, BitVector, obj)
+
+## Ranges
+gap_to_julia_intern(::Type{T}, obj::GapObj, recursive::Bool, dict::Union{Nothing,RecDict}) where {T<:UnitRange} = _get!(dict, T, obj)
+gap_to_julia_intern(::Type{T}, obj::GapObj, recursive::Bool, dict::Union{Nothing,RecDict}) where {T<:StepRange} = _get!(dict, T, obj)
+
+
+#
+# The rules for caching/duplicate tracking during conversion from GAP to Julia,
+# based on the conversion target type `T`
+# - if `T` is a bits type then identity and equality coincide, so no tracking is needed
+# - if `T` is `GapObj` then the only "conversion" we permit is the identity map, so
+#   no tracking is needed
+# - if `T` is a container type (e.g. `Vector`) then the conversion code in general
+#   ought to first add the conversion result to the tracking dictionary before
+#   performing any recursive calls.
+#
+# At the top level of conversion, the tracking dict may be set to `nothing`.
+# Then the conversion method has two ways to proceed: If `T` is *not* a
+# container type then `dict` can simply be ignored.
+# If `T` is a container type, then in general `dict` needs to be initialized
+# and subobjects tracked in it (but not the top level container), unless the
+# type of the container elements does not require tracking, in which case
+# again `dict` can be simply ignored.
+#
+
+# default: bits types don't need tracking, all other types do
+_needs_conversion_tracking(::Type{T}) where T = !isbitstype(T)
+
+# the only way conversion produces a GapObj is if the input was a GapObj, and
+# then we just use the same GapObj, so duplicates are automatically taken care of
+# and we don't need tracking
+_needs_conversion_tracking(::Type{GapObj}) = false
+
+# Union types need tracking if any of the types the union was formed over need it
+#_needs_conversion_tracking(T::Union) = any(_needs_conversion_tracking, Base.uniontypes(T))
+# TODO: alternatively just handle GAP.GapInt, GAP.Obj
```

Comment:
As ar as I understand, in line 46 above, it is stated that union types are not allowed.